### PR TITLE
FlyteCoPilot: Raw container support [Alpha]

### PR DESCRIPTION
**Pull request by kumare3**
# TL;DR
This change adds a new type of task called SDKRawContainer task. Currently this is in Alpha, but eventually will become the base task for all container executions on K8s.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
Refer to the Design doc at https://docs.google.com/document/d/1ZsCDOZ5ZJBPWzCNc45FhNtYQOxYHz0PAu9lrtDVnUpw/edit?ts=5ea7f579#

## Tracking Issue
https://github.com/lyft/flyte/issues/297

## Follow-up issue
NA

### CODE REVIEW

File: flytekit/common/tasks/raw_container.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+        :param discoverable:
+        :param discovery_version:
+        :param retries:
+        :param timeout:
+        :param input_data_dir: This is the directory where data will be downloaded to
+        :param output_data_dir: This is the directory where data will be uploaded from
+        :param metadata_format: Format in which the metadata will be available for the script
+        """
+
+        # Set as class fields which are used down below to configure implicit
```
Comment (wild-endeavor) [inline flytekit/common/tasks/raw_container.py:181]: these aren't class fields though.

File: flytekit/common/tasks/raw_container.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+                _task_models.RuntimeMetadata.RuntimeType.FLYTE_SDK, __version__,
+                "python"),
+            timeout or _datetime.timedelta(seconds=0),
+            _literals.RetryStrategy(retries),
+            interruptible,
+            discovery_version,
+            "This is deprecated!"
+        )
+
+        # Here we set the routing_group, catalog, and schema as implicit
```
Comment (wild-endeavor) [inline flytekit/common/tasks/raw_container.py:202]: leftover comment from presto?

File: flytekit/common/tasks/raw_container.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+        # parameters for caching purposes
+        i = _interface.TypedInterface(inputs=types_to_variable(inputs), outputs=types_to_variable(outputs))
+
+        # TODO create custom proto to store data dir and other things
```
Comment (wild-endeavor) [inline flytekit/common/tasks/raw_container.py:206]: remove

File: tests/flytekit/unit/common_tests/tasks/test_raw_container_task.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
+from flytekit.common.tasks.raw_container import SdkRawContainerTask
+from flytekit.sdk.types import Types
+
+
+def test_raw_container_task_definition():
```
Comment (wild-endeavor) [inline tests/flytekit/unit/common_tests/tasks/test_raw_container_task.py:5]: can you do a test in your test harness, and maybe add one here, where you have a workflow with two of these raw container tasks and the second one consumes the output of the first one?  Or is that not possible to do?

File: flytekit/common/tasks/raw_container.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+        args=args,
+        resources=_task_models.Resources(limits=limits, requests=requests),
+        env=environment,
+        config={}
+    )
+
+
+class SdkRawContainerTask(_base_task.SdkTask):
+    """
+    This class includes the logic for building a task that executes as a Presto task.
```
Comment (wild-endeavor) [inline flytekit/common/tasks/raw_container.py:123]: Can we add a few paragraphs of documentation here or somewhere else on 1) how to use this and 2) how it works?

For 1) I've left you a page to fill in here: https://github.com/lyft/flytesnacks/blob/cookbook/cookbook/workflows/recipe_5/index.rst

Feel free to check out that branch and commit... or make your own PR if you want the street cred.

File: tests/flytekit/unit/common_tests/tasks/test_raw_container_task.py (hunk 4 above)
Comment (kumare3) [reply tests/flytekit/unit/common_tests/tasks/test_raw_container_task.py:5]: it is possible, good point. I actually have a real workflow, just scroll up. But i will write an test workflow too

File: flytekit/common/tasks/raw_container.py (hunk 3 above)
Comment (kumare3) [reply flytekit/common/tasks/raw_container.py:206]: no not remove, add the custom proto I guess?

File: flytekit/common/tasks/raw_container.py (hunk 2 above)
Comment (kumare3) [reply flytekit/common/tasks/raw_container.py:202]: true!

File: flytekit/common/tasks/raw_container.py (hunk 1 above)
Comment (kumare3) [reply flytekit/common/tasks/raw_container.py:181]: `      self.input_data_dir = input_data_dir
        self.output_data_dir = output_data_dir
        self.metadata_format = metadata_format`

File: flytekit/common/tasks/raw_container.py (hunk 5 above)
Comment (kumare3) [reply flytekit/common/tasks/raw_container.py:123]: nice :D will do

File: tests/flytekit/common/workflows/raw_container.py [hunk 6]
```diff
@@ -10,14 +10,14 @@
...
-    command=["sh", "-c", "cd /var/flyte/; paste ./inputs/val | awk '{print ($1 * $1)}' > ./outputs/out"],
+    command=["sh", "-c", "cd /var/; paste ./inputs/val | awk '{print ($1 * $1)}' > ./outputs/out"],
 )
 
 echo = SdkRawContainerTask(
     input_data_dir="/var/flyte/inputs",
     inputs={"x": Types.Integer},
     image="alpine",
-    command=["cat", "/var/flyte/inputs.json"],
+    command=["sh" "-c", "ls /var/flyte/inputs; cat /var/flyte/inputs/inputs.json"],
```
Comment (wild-endeavor) [inline tests/flytekit/common/workflows/raw_container.py]: why not make this a bit more complicated?  `echo '3 + 5' | bc > output_file`. where 3 and or 5 comes from an input?

File: tests/flytekit/unit/common_tests/tasks/test_raw_container_task.py (hunk 4 above)
Comment (kumare3) [reply tests/flytekit/unit/common_tests/tasks/test_raw_container_task.py:5]: actually i cannot write a test today, as it will execute it locally, which we do not want

File: tests/flytekit/common/workflows/raw_container.py (hunk 6 above)
Comment (kumare3) [reply tests/flytekit/common/workflows/raw_container.py]: the previous step is complicated

Comment (kumare3) [reply tests/flytekit/common/workflows/raw_container.py]: if you want i can modify this to produce an output?

File: flytekit/common/tasks/raw_container.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+                storage_limit=storage_limit,
+                cpu_limit=cpu_limit,
+                gpu_limit=gpu_limit,
+                memory_limit=memory_limit,
+                environment=environment,
+            )
+        )
+
+    # Override method in order to set the implicit inputs
+    def __call__(self, *args, **kwargs):
```
Comment (wild-endeavor) [inline flytekit/common/tasks/raw_container.py:232]: Is this necessary?

**Comment by wild-endeavor** [comment]
Also, bump version.

**Comment by wild-endeavor** [comment]
Also, it looks like python 2.7 tests are failing.  let's just stop running them.  it's in travis.yml.  I thought we had done that already but i guess not.

File: flytekit/__init__.py [hunk 8]
```diff
@@ -2,4 +2,4 @@
 
 import flytekit.plugins
 
-__version__ = '0.9.1'
+__version__ = '0.9.3'
```
Comment (wild-endeavor) [inline flytekit/__init__.py:5]: i made this 0.9.3 cuz I thought katrina's change for single-task would go in first but I think this will go in first actually.
